notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+    "        valid, \n",
+    "        prediction, \n",
+    "        **HEADER\n",
+    "    )\n",
+    "    \n",
+    "    rmse = rating_eval.rmse()\n",
+    "    \n",
+    "    # Return the objective function result.\n",
+    "    return {\n",
+    "       'loss': rmse,\n",
+    "       'status': STATUS_OK,\n",
+    "       'eval_time': time.time()\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A search space is usually defined for hyperparameter exploration. Design of search space is empirical, and depends on the understanding of how distribution of parameter of interest affects the model performance measured by the loss function. \n",
+    "\n",
+    "In the ALS algorithm, the two hyper parameters, rank and reg, affect model performance in a way that\n",
+    "* The higher the rank, the better the model performance but also the higher risk of overfitting.\n",
+    "* The reg parameter is subject to rank and it prevents overfitting in certain way. \n",
+    "\n",
+    "Therefore, in this case, a uniform distribution and a lognormal distribution sampling spaces are used for rank and reg, respectively. A narrow search space is used for illustration purpose, that is, the range of rank is from 10 to 20, while that of reg is from $e^{-5}$ to $e^{-1}$. Together with the randomly sampled hyper parameters, other parameters use for building / evaluating the recommender, like `k`, column names, data, etc., are kept as constants."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define a search space\n",
+    "space = {\n",
+    "    'rank': hp.quniform('rank', 10, 20, 5),\n",
+    "    'reg': hp.loguniform('reg', -5, -1),\n",
+    "    'train': train, \n",
+    "    'valid': valid, \n",
+    "    'col_user': \"UserId\", \n",
+    "    'col_item': \"MovieId\", \n",
+    "    'col_rating': \"Rating\", \n",
+    "    'col_prediction': \"prediction\", \n",
+    "    'k': 10,\n",
+    "    'relevancy_method': \"top_k\"\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`fmin` of `hyperopt` is used for running the trials for searching optimal hyper parameters. In `hyperopt`, there are different strategies for intelligently optimize hyper parameters. For example, `hyperopt` avails Tree of Parzen Estimators (TPE) method for searching optimal parameters, which is more efficient than random search or grid search (details can be found [here](http://hyperopt.github.io/hyperopt/)). It is claimed in the project website that Bayesian based optimization method is planned but so far it has not yet been implemented. \n",
+    "\n",
+    "The following runs the trials with the pre-defined objective function and search space. TPE is used as the optimization method. Totally there will be 10 evaluations run for searching the best parameters."
```

Comment:
In `details can be found [here](http://hyperopt.github.io/hyperopt/)` we are linking to the main page of hyperopt. I couldn't find a description of TPE on their page, which is odd. Maybe we can do a brief description of TPE and link to some source like wikipedia or other?

Reply:
Added reference.